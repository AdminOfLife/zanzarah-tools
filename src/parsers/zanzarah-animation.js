const { DynamicArray } = require('../utils/jbinary');

exports.typeSet = {
  'jBinary.all': 'Animation',
  'jBinary.littleEndian': true,

  DynamicArray,

  Animation: {
    keyFrameCount: 'uint32',
    flags: 'int32',
    duration: 'float32',
    keyFrames: ['array', 'KeyFrame', 'keyFrameCount'],
  },

  KeyFrame: {
    rotation: ['array', 'float32', 4],
    position: ['array', 'float32', 3],
    time: 'float32',
    parent: 'uint32',
  },
};
