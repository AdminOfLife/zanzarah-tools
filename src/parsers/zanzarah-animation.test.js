import jBinary from 'jbinary';
import { typeSet } from './zanzarah-animation';

const parse = async fileNameOrBuffer => {
  let binary = null;
  if (Buffer.isBuffer(fileNameOrBuffer)) {
    binary = new jBinary(fileNameOrBuffer, typeSet);
  } else {
    binary = await jBinary.load(fileNameOrBuffer, typeSet);
  }
  return binary.readAll();
};

describe('parsers - zanzarah animation', () => {
  it('exports typeSet to parse', () => {
    expect(typeSet).toBeDefined();
  });
  pit('parses smallest known files', async () => {
    expect(await parse('./public/fixtures/smallest-animation.ska')).toMatchSnapshot();
  });
  pit('parses some random file', async () => {
    expect(await parse('./public/fixtures/animation.ska')).toMatchSnapshot();
  });
});
