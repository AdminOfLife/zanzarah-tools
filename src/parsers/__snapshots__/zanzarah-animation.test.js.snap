exports[`parsers - zanzarah animation parses smallest known files 1`] = `
Object {
  "duration": 6.066666603088379,
  "flags": 0,
  "keyFrameCount": 2,
  "keyFrames": Array [
    Object {
      "parent": 4147921032,
      "position": Array [
        -3.5539347464919047e-8,
        1.5900786465294914e-8,
        4.943495213183269e-8,
      ],
      "rotation": Array [
        0,
        0,
        0,
        -1,
      ],
      "time": 0,
    },
    Object {
      "parent": 0,
      "position": Array [
        -3.5539347464919047e-8,
        1.5900786465294914e-8,
        4.943495213183269e-8,
      ],
      "rotation": Array [
        0,
        0,
        0,
        -1,
      ],
      "time": 6.066666603088379,
    },
  ],
}
`;

exports[`parsers - zanzarah animation parses some random file 1`] = `
Object {
  "duration": 0.6399999856948853,
  "flags": 0,
  "keyFrameCount": 28,
  "keyFrames": Array [
    Object {
      "parent": 4048534608,
      "position": Array [
        -4.55869508808604e-10,
        5.6764194039660154e-11,
        -1.073859312583636e-8,
      ],
      "rotation": Array [
        0,
        0,
        0,
        -1,
      ],
      "time": 0,
    },
    Object {
      "parent": 4048534608,
      "position": Array [
        0.014641770161688328,
        -6.714202527291491e-8,
        1.9936447159807358e-8,
      ],
      "rotation": Array [
        0.7071067690849304,
        -6.280369833130267e-15,
        0.7071067690849304,
        -4.214685134229512e-8,
      ],
      "time": 0,
    },
    Object {
      "parent": 4048534608,
      "position": Array [
        -0.008825178258121014,
        0.0165800042450428,
        0.0109342560172081,
      ],
      "rotation": Array [
        1.1530437164824136e-16,
        0,
        5.765230229115093e-16,
        -1,
      ],
      "time": 0,
    },
    Object {
      "parent": 4048534608,
      "position": Array [
        -0.37488803267478943,
        0.1454269289970398,
        0.04480353742837906,
      ],
      "rotation": Array [
        -0.03006048873066902,
        0.03006049059331417,
        -0.25193148851394653,
        -0.9668580293655396,
      ],
      "time": 0,
    },
    Object {
      "parent": 4048534608,
      "position": Array [
        0.008812627755105495,
        0.01658000238239765,
        0.010903159156441689,
      ],
      "rotation": Array [
        1.1530437164824136e-16,
        0,
        5.765230229115093e-16,
        -1,
      ],
      "time": 0,
    },
    Object {
      "parent": 4048534608,
      "position": Array [
        0.37597745656967163,
        0.1454269289970398,
        0.04483472555875778,
      ],
      "rotation": Array [
        -0.039559636265039444,
        -0.03319445624947548,
        0.26413804292678833,
        -0.9630629420280457,
      ],
      "time": 0,
    },
    Object {
      "parent": 0,
      "position": Array [
        -4.55869508808604e-10,
        5.6764194039660154e-11,
        -1.073859312583636e-8,
      ],
      "rotation": Array [
        0,
        0,
        0,
        -1,
      ],
      "time": 0.6399999856948853,
    },
    Object {
      "parent": 36,
      "position": Array [
        0.014641770161688328,
        -6.714202527291491e-8,
        1.9936447159807358e-8,
      ],
      "rotation": Array [
        0.7071067690849304,
        -6.280369833130267e-15,
        0.7071067690849304,
        -4.214685134229512e-8,
      ],
      "time": 0.6399999856948853,
    },
    Object {
      "parent": 72,
      "position": Array [
        -0.008825178258121014,
        0.0165800042450428,
        0.0109342560172081,
      ],
      "rotation": Array [
        1.3015241107083853e-9,
        -1.8292981573964084e-9,
        -0.07254859060049057,
        -0.9973108172416687,
      ],
      "time": 0.03999999910593033,
    },
    Object {
      "parent": 108,
      "position": Array [
        -0.37488803267478943,
        0.1454269289970398,
        0.04480353742837906,
      ],
      "rotation": Array [
        -0.03006049059331417,
        0.03006049059331417,
        -0.25193145871162415,
        -0.9668580293655396,
      ],
      "time": 0.6399999856948853,
    },
    Object {
      "parent": 144,
      "position": Array [
        0.008812627755105495,
        0.01658000238239765,
        0.010903159156441689,
      ],
      "rotation": Array [
        -1.3242434926397095e-9,
        2.3852082531305996e-9,
        0.07254859805107117,
        -0.9973108172416687,
      ],
      "time": 0.03999999910593033,
    },
    Object {
      "parent": 180,
      "position": Array [
        0.37597745656967163,
        0.1454269289970398,
        0.04483472928404808,
      ],
      "rotation": Array [
        -0.039559636265039444,
        -0.03319445624947548,
        0.26413804292678833,
        -0.9630629420280457,
      ],
      "time": 0.6399999856948853,
    },
    Object {
      "parent": 288,
      "position": Array [
        -0.008825178258121014,
        0.0165800042450428,
        0.0109342560172081,
      ],
      "rotation": Array [
        2.8221414094531383e-9,
        -3.1825038160349095e-9,
        -0.15481474995613098,
        -0.9879646897315979,
      ],
      "time": 0.07999999821186066,
    },
    Object {
      "parent": 360,
      "position": Array [
        0.008812627755105495,
        0.01658000238239765,
        0.010903159156441689,
      ],
      "rotation": Array [
        -2.8996902656786006e-9,
        5.670639513510878e-9,
        0.15481474995613098,
        -0.9879646897315979,
      ],
      "time": 0.07999999821186066,
    },
    Object {
      "parent": 432,
      "position": Array [
        -0.008825178258121014,
        0.0165800042450428,
        0.0109342560172081,
      ],
      "rotation": Array [
        8.72413430386132e-9,
        -1.1466764204826063e-9,
        -0.48665106296539307,
        -0.8736038208007812,
      ],
      "time": 0.2800000011920929,
    },
    Object {
      "parent": 468,
      "position": Array [
        0.008812627755105495,
        0.01658000238239765,
        0.010903159156441689,
      ],
      "rotation": Array [
        -8.872375723001369e-9,
        2.7144015035673874e-8,
        0.48665106296539307,
        -0.8736038208007812,
      ],
      "time": 0.2800000011920929,
    },
    Object {
      "parent": 504,
      "position": Array [
        -0.008825178258121014,
        0.0165800042450428,
        0.0109342560172081,
      ],
      "rotation": Array [
        8.728896716547752e-9,
        -1.2583416530986824e-9,
        -0.5,
        -0.8660253882408142,
      ],
      "time": 0.3199999928474426,
    },
    Object {
      "parent": 540,
      "position": Array [
        0.008812627755105495,
        0.01658000238239765,
        0.010903159156441689,
      ],
      "rotation": Array [
        -8.728894940190912e-9,
        2.8543974295303087e-8,
        0.5,
        -0.8660253882408142,
      ],
      "time": 0.3199999928474426,
    },
    Object {
      "parent": 576,
      "position": Array [
        -0.008825178258121014,
        0.0165800042450428,
        0.0109342560172081,
      ],
      "rotation": Array [
        8.04110644736511e-9,
        -2.7308690864202845e-9,
        -0.4826834499835968,
        -0.8758366703987122,
      ],
      "time": 0.36000001430511475,
    },
    Object {
      "parent": 612,
      "position": Array [
        0.008812627755105495,
        0.01658000238239765,
        0.010903159156441689,
      ],
      "rotation": Array [
        -7.790775136129469e-9,
        2.7248331591067654e-8,
        0.4826834499835968,
        -0.8758366703987122,
      ],
      "time": 0.36000001430511475,
    },
    Object {
      "parent": 648,
      "position": Array [
        -0.008825178258121014,
        0.0165800042450428,
        0.0109342560172081,
      ],
      "rotation": Array [
        6.675413555257137e-9,
        -5.228563804138275e-9,
        -0.4352594017982483,
        -0.9003037810325623,
      ],
      "time": 0.4000000059604645,
    },
    Object {
      "parent": 684,
      "position": Array [
        0.008812627755105495,
        0.01658000238239765,
        0.010903159156441689,
      ],
      "rotation": Array [
        -6.0758602593580235e-9,
        2.361062456657237e-8,
        0.4352594316005707,
        -0.9003037810325623,
      ],
      "time": 0.4000000059604645,
    },
    Object {
      "parent": 720,
      "position": Array [
        -0.008825178258121014,
        0.0165800042450428,
        0.0109342560172081,
      ],
      "rotation": Array [
        -2.131884224709779e-9,
        -1.0741092459909396e-8,
        -0.08386237919330597,
        -0.9964537024497986,
      ],
      "time": 0.5600000023841858,
    },
    Object {
      "parent": 756,
      "position": Array [
        0.008812627755105495,
        0.01658000238239765,
        0.010903159156441689,
      ],
      "rotation": Array [
        4.4670440679794865e-9,
        5.5357913808506964e-9,
        0.08386236429214478,
        -0.9964537024497986,
      ],
      "time": 0.5600000023841858,
    },
    Object {
      "parent": 792,
      "position": Array [
        -0.008825178258121014,
        0.0165800042450428,
        0.0109342560172081,
      ],
      "rotation": Array [
        -4.178324353176777e-9,
        -9.964178815380365e-9,
        8.218991503686603e-16,
        -1,
      ],
      "time": 0.6000000238418579,
    },
    Object {
      "parent": 828,
      "position": Array [
        0.008812627755105495,
        0.01658000238239765,
        0.010903159156441689,
      ],
      "rotation": Array [
        6.891679671383599e-9,
        3.1407214606815614e-9,
        5.679968951435659e-16,
        -1,
      ],
      "time": 0.6000000238418579,
    },
    Object {
      "parent": 864,
      "position": Array [
        -0.008825178258121014,
        0.0165800042450428,
        0.0109342560172081,
      ],
      "rotation": Array [
        -4.178324353176777e-9,
        -9.964178815380365e-9,
        8.218991503686603e-16,
        -1,
      ],
      "time": 0.6399999856948853,
    },
    Object {
      "parent": 900,
      "position": Array [
        0.008812627755105495,
        0.01658000238239765,
        0.010903159156441689,
      ],
      "rotation": Array [
        6.891679671383599e-9,
        3.1407214606815614e-9,
        5.679968951435659e-16,
        -1,
      ],
      "time": 0.6399999856948853,
    },
  ],
}
`;
